rfcs/2024-05-21-ui-component-model.md

```diff
@@ -0,0 +1,140 @@
+    - Allows the runtime to enforce data policies on component
+- Components have **local state** (P1)
+    - State is encapsulated within component
+    - Components may pass state down to child components as input
+        - E.g. the React **[“lifting state up” pattern](https://legacy.reactjs.org/docs/lifting-state-up.html)**.
+    - Local state may be **persisted**.
+        - If it isn’t, it is **ephemeral**, and lasts for the lifetime of the component.
+- Components are **islands** (P1)
+    - Components can be used free-standing, or within a larger component tree.
+    - Note: this is in contrast to something like the Elm App Architecture Pattern, where models, views, and update functions are “zippered” together, meaning components are “some assembly required”. This would fall short of this goal without some additional means of making an individual component a free-standing island.
+- Components are **composable** (P1)
+    - Components can be combined together like lego to create larger components
+    - Composing plugins should be as easy as plugging together component inputs, outputs, and events, and arranging a UI tree. It shouldn’t be more complicated than that.
+- Components can have **holes**, allowing you to slot in an arbitrary component of the right shape. (P1)
+    - Inversion of control for templates.
+    - The shape of the hole is determined by the data’s input and output types
+    - Example mechanisms
+        - [slots](https://developer.mozilla.org/en-US/docs/Web/API/Web_components/Using_templates_and_slots)
+        - [passing down closures that evaluate to components](https://swiftwithmajid.com/2019/11/06/the-power-of-closures-in-swiftui/) 
+        - [overridable blocks](https://mustache.github.io/mustache.5.html#Blocks)
+- Components have **[high locality of behavior](https://github.com/gordonbrander/generative-ui-playbook?tab=readme-ov-file#llms-work-best-with-high-locality)** (P1)
+    - All component behavior is colocated, including structure, style, and behavior
+- Components are relatively **small**
+    - LLMs will have higher accuracy generating small islands of interactivity vs whole apps
+    - Small composable components are easier to understand and debug
+- UI **templates** are **pure functions** (P1)
+    - Templates take inputs, and output a UI tree and events
+    - Templates produce a UI tree that is easy for the runtime to analyze and sanitize (probably a VDOM, probably not raw DOM).
+- Components are renderable to web (P1)
+    - Other platforms may be supported in future, but web platform is primary
+
+Soft goals:
+
+- UI templates are static (P3)
+    - They are compiled once at program start, and produce a static tree with specific “binding points” in the tree, where dynamic values and dynamic lists are rendered.
+    - This may have a performance advantage over a totally dynamic VDOM tree, since it would allow us to analyze and enforce policies on the tree once, rather than after every render
```

Comment:
Throwing it out there: if we know the types for all inputs of a control we could use a "fuzzer" approach for static policy validation even with dynamic subtrees, try a huge range of inputs etc. and ensure the output looks "safe".

Reply:
My understanding is (static) policy is enforced at RecipeModule boundaries by the Runtime, not within a JavaScript Context/CodeModule